### ISSUE TITLE
Bcbio 1.1.3: RNAseq variants soft-filtering is too strict 

### ISSUE THREAD

**Issue by WimSpee** [bcbio/bcbio-nextgen#2683]
Hi, 

RNAseq variants soft-filtering in bcbio 1.1.3 seem to be too strict.

We used picard genotype concordance to compare the (bcbio 1.1.2) RNAseq and WGRS based genotypes of the same sample. 
RNAseq = STAR + GATK4 GVCF
WGRS = BWA + GATK4 GVCF

A sankey plot of the genotype concordance shows that:

- c.a. 1/3 of the RNAseq filtered variants are true HOM_VAR variants in the WGRS data (RNAseq filter is wrong)
- c.a. 1/3 of the RNAseq filtered variants are true HET_VAR variants in the WGRS data (RNAseq filter is wrong)
- c.a. 1/3 of the RNAseq filtered variants are true MISSING variants in the WGRS data (RNAseq filter is correct)

![rna_seq_soft_filter_bcbio_1 1 2_gt_concordance_v s _wgrs](https://user-images.githubusercontent.com/34706930/52947950-73004380-3378-11e9-8b6b-dbac342b662a.jpg)

I thought this might have been fixed in bcbio 1.1.3. See this issue and this statement from the history log
https://github.com/bcbio/bcbio-nextgen/issues/1502

> RNA-seq variation: fix for over-filtering variants near splice junctions with STAR.

So I reran the RNAseq analysis with bcbio 1.1.3 in the hope that a higher fraction of the RNAseq filtered variants would be correctly filtered. 

However the genotype concordance and PPV after rerunning the RNAseq in bcbio 1.1.3 is slighty worse.
And the sankey plot still looks the same. 

![rna_seq_soft_filter_bcbio_1 1 3_gt_concordance_v s _wgrs](https://user-images.githubusercontent.com/34706930/52950339-20765580-337f-11e9-9f8a-724752ae190f.jpg)

Do you see the same thing when comparing RNAseq data against WGRS data in your benchmark(s)?

Could you share on what the RNAseq filter method is currently based? 
And if there are maybe ways to improve the specificity of this filter?
 (i.e. higher fraction of RNA_SEQ_VC_FILTERD that map to WGRS_MISSING?  )

Thank you.

**Comment by roryk** [comment]
H @WimSpee,

First: thank you for looking at this, variant calling with RNA-seq is definitely an area where we can do better in bcbio.

@naumenko-sa has way more experience looking at this than I do, so I'm pinging him. A bit ago he identified that GATK4 is doing poorly on RNA-seq variants in #2410-- it might be worth switching to GATK3.

The filtering we do is in two parts, the first is based on here: https://software.broadinstitute.org/gatk/documentation/article.php?id=3891

The second is implemented in the issue you linked, we also filter at splice junctions. The reason why is I took Sergey's truth dataset from here: https://github.com/bcbio/bcbio-nextgen/issues/1502 and looked at where the false positives lie.

![image](https://user-images.githubusercontent.com/414586/52966198-b193ff00-3374-11e9-8f4c-417f515bb94c.png)

You can see that there are very many more false positives right around the splice junctions, so we remove those, even if it means we are missing important donor-acceptor type variants. We just can't call accurately there:

![image](https://user-images.githubusercontent.com/414586/52966286-e56f2480-3374-11e9-8fea-5529f344fe44.png)

I am very happy you are looking carefully at this, and would like to work to make this better in bcbio. We're totally open to whatever suggestions you may have to make it better.

**Comment by naumenko-sa** [comment]
Hi, @WimSpee  and @roryk!

Thanks for inviting me to this discussion! Calling variants in RNA-seq seems very challenging, and very few people use it. I saw that variants around junctions were filtered out recently, and I was very curious about that. I have not tested it yet by myself.

@WimSpee:
1. What is WGRS? Is it WGS or something else? Are you using NA12878 to validate or it is an internal sample WGS + RNAseq? Could you please provide coverage info for WGS and RNAseq? Tissue for RNA-seq?
2. Could you please try tools_off: gatk4? Is @roryk mentioned, I found gatk4 calls in RNA-seq problematic, and since that I'm using gatk3.8 for RNA-seq.
3. I saw a very big difference between SNPs and Indels in RNAseq: for SNPs we may achieve 95% precision, Indels are not reliable at all. Could you please plot separate figures for SNPs and Indels, if possible?

@roryk:
I'm don't have any suggestions about variant filtration near splice junctions for now. I'm going to test it to compare with the previous results. I think there are two different use cases for variant calling in RNAseq: 
1) when there is no WES or WGS, and RNA-seq is the only source of data to call variants. In that case (probably, the default one), variant filtration near junctions makes total sense (and also recommended by famous Piskol2013 article). 
2) A hunt for splicing affecting variants, not captured by WES or where WES is not available. Here variants near junctions are needed (probably, only SNPs). Or they could just remain in the vcf as filtered out by the JUNCTION_SITE filter. Usually, search for such variants starts not from the variant list itself (too many false positives to interpret), but from the functional evidence (aberrant splicing found, expression level changed), and it is important to trace the causal variant if we know the gene, or the locus. Sometimes, there are brilliants in that trash of FP variants.

Sergey

**Comment by WimSpee** [comment]
Hi @roryK and @naumenko-sa. 

Thank you for the information. 
I am starting to apprecicate (more) that it is difficult to do sensitive and specific filtering for RNAseq discovered variants. Especially close to the splice junctions. 

WGRS = Whole Genome ReSequencing = our internal term for WGS not for denovo assembly

Where it is cost-effective we use WGRS for variant discovery.  
Variant discovery via RNAseq is for us:
* a free extra when we generate / use RNAseq data for other purposes 
* one of the alternatives for variant discovery where WGRS is not (yet) cost effective

For these two purposes we are happy already with the current support in bcbio for RNAseq variant calling via GATK4. 

I am trying to figure out now how reliable RNAseq is for variant discovery.
For this I generated WGRS and RNAseq data of the same samples. This is not a standard thing that we do. 

Species is Lactuca sativa, tissue is leaves. 
RNAseq is 47M to 60M 100bp reads per sample. Medium to high coverage, don't know the exact coverage number. 
WGRS = 13X to 18X coverage. The samples are mostly homozygous, therefore 30X is not needed.  

The Positive Predictive Value of the PASS HOM_VAR RNAseq SNPs c.a. 90% v.s. the WGRS data. This was a positive suprise. 
I was just also surprised by how many FILTERED HOM_VAR RNAseq SNPs are PASS in the WGRS data. 

We are not looking at indels in the RNAseq data. We also found that they are much less reliable. And we prefer to use SNPs anyway. 
The figures from the issue opening comments are also just for SNPs. 

It would be nice if the  RNAseq variant filtering could be more specific.

Just understanding more about how and why the RNAseq variant filtering works like it does is already useful. Like @naumenko-sa  said, it is useful to know there can be gems in the RNAseq filtered variants. 
Thank you again for that information.

We don't have a license for GATK3. 
It also makes more sense to me to have RNAseq variant calling and filtering further improved in GATK4 (here or at the Broad). 

I'll have a look if I can find anything that distinguishes between the 3 RNAseq filtered variant categories 

- HET concordant with WGRS
- HOM concordant WGRS, 
- HET/HOM discordant with WGRS

I'll let you know here if I can or can't find something. 

Thanks again for your information.

**Comment by WimSpee** [comment]
Hi @roryk  It turns out that almost all of the filtered variants are filtered because of the SnpCluster filter. 
There is not much difference to where the SNP are located in relation to the splice sites.

![variant_category_per_nearest_splice_site_distance](https://user-images.githubusercontent.com/34706930/53558224-d98b1b80-3b47-11e9-98a2-5407168ed8c0.png)

Most of these RNAseq SnpCluster variants (bottom track)  are reals SNPs in whole genome re-sequencing (top track) .

![vsdev-1360_snp_cluster_1_clean](https://user-images.githubusercontent.com/34706930/53558285-fe7f8e80-3b47-11e9-9aa1-56a8da6a5759.png)

![vsdev-1360_snp_cluster_2_clean](https://user-images.githubusercontent.com/34706930/53558298-09d2ba00-3b48-11e9-88bf-143da9c73f07.png)

I suspect that the current SnpCluster cutoff values make most sense in Human, where these SnpCluster SNPs are less likely to represent true SNPs but to be artifacts. The genetic diversity in the species that we work with is higher, and the reference genome maybe also more distant to the re-sequenced samples. 

Would it be possible in bcbio to add via the analysis yaml:

1.  a switch for the RNAseq variant calling in bcbio to not add SnpCluster annotation?
2.  be able so set custom cutoff values for the SnpCluster annotation? i.e. X SNPs in Y bases. 

Thank you.

**Comment by naumenko-sa** [comment]
Hi @WimSpee!
Thanks for sharing this data! Yes, I think you found the main point: polymorphism levels (genetic diversity) in a population. For _H.sapiens_ it is 0.1% and almost all of the variants in SNP clusters are FP. What is the genetic diversity of _L.sativa_?
Sergey

**Comment by WimSpee** [comment]
The SNP frequency depends very much on if you include some / all of the wild relatives of L.sativa, that you can still cross with L.sativa. 
If you include all the wild relatives, then it is very much possible to get up to and even beyond 1 small variant every 10 bases, against the reference genome of L.sativa. 

If you exclude all the wild relatives, then the SNP frequency is much lower.
So we have to be more careful with SnpClusters, and also reference bias is a big issue for us. Some samples are from mixed backgrounds, so it is not easy to pick a single best reference genome that is equally useful  for all samples. 

See also here for some more general backgound context:
http://home.cc.umanitoba.ca/~frist/Seminars/biotech/widecross.html

**Comment by WimSpee** [comment]
We experimented with some different SnpCluster cutoff values, by overwriting these here in the bcbio code:
https://github.com/bcbio/bcbio-nextgen/blob/fc610d63b0bd27e5d9adc07ae069fc5db4cb35ae/bcbio/rnaseq/variation.py#171

None of the cutoff values that we tried showed a clear improvement over the default 3 in 35 values. 
We always were left with a big mixed bag of SNPs, and we rather have this big mixed bag of SNP variants labeled with SnpCluster than with PASS. 

So never mind the request to disable or to allow custom values. 
Thank you for you thoughts and help on this issue.
